File: frontend/src/interfaces/bot.ts
import { LLMEngine, ImageSize, GPT35Tokens, GPT4Tokens } from '@/interfaces/enums';

export interface AiSettingsSchema {
  ai_goals: string[];
  ai_name: string;
  ai_role: string;
  api_budget?: number | null;
}

export interface BaseBotSchema {
  smart_tokens: GPT4Tokens;
  fast_tokens: GPT35Tokens;
  smart_engine: LLMEngine;
  fast_engine: LLMEngine;
  image_size: ImageSize;
  ai_settings: AiSettingsSchema | null;
}

export interface BotInCreateSchema extends BaseBotSchema {
  ai_settings: AiSettingsSchema;
}

export interface BotSchema extends BaseBotSchema {
  ai_settings: AiSettingsSchema;
  is_active: boolean;
  is_failed: boolean;
  runs_left: number;
}

export interface BotFormSchema extends BaseBotSchema {
  ai_goal: string;
}

export interface WorkspaceFileSchema {
  name: string;
  path: string;
  is_dir: boolean;
  mime_type: string | null;
  size: string | null;
}

export interface WorkspaceFileEnrichedSchema extends WorkspaceFileSchema {
  show: boolean;
}

File: backend/src/app/schemas/bot.py
from typing import Any

from pydantic import BaseModel, Field
from prisma.partials import BotInCreateSchema as _BotInCreateSchema, BotSchema as _BotSchema

from app.schemas.enums import ImageSize, EngineTokens, LLMEngine


class AiSettingsSchema(BaseModel):
    ai_goals: list[str] = Field(..., max_items=20, max_length=2000)
    ai_name: str = Field(..., max_length=30)
    ai_role: str = Field(..., max_length=1200)
    api_budget: float = Field(0.0, ge=0)


class BotSchema(_BotSchema):
    ai_settings: dict[str, Any]


class BotInCreateSchema(_BotInCreateSchema):
    fast_engine: LLMEngine = LLMEngine.GPT_3_5_TURBO
    smart_engine: LLMEngine = LLMEngine.GPT_4
    fast_tokens: EngineTokens = EngineTokens.t4000
    smart_tokens: EngineTokens = EngineTokens.t4000
    image_size: ImageSize = ImageSize.s512
    ai_settings: AiSettingsSchema


class WorkspaceFileSchema(BaseModel):
    name: str
    path: str
    is_dir: bool
    mime_type: str | None = None
    size: str | None = None
